services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -510,40 +515,32 @@ public static void dettachClientSession(UserSessionProvider sessions, RealmModel
 
 
     /** Return client itself + all default client scopes of client + optional client scopes requested by scope parameter **/
-    public static Set<ClientScopeModel> getRequestedClientScopes(String scopeParam, ClientModel client) {
-        // Add all default client scopes automatically
-        Set<ClientScopeModel> clientScopes = new HashSet<>(client.getClientScopes(true, true).values());
-
-        // Add client itself
-        clientScopes.add(client);
+    public static Stream<ClientScopeModel> getRequestedClientScopes(String scopeParam, ClientModel client) {
+        // Add all default client scopes automatically and client itself
+        Stream<ClientScopeModel> clientScopes = Stream.concat(
+                client.getClientScopes(true, true).values().stream(),
+                Stream.of(client)).distinct();
```

Comment:
Shouldn't distinct() be placed in the stream that is returned in line 531 bellow?

Reply:
Yeah but probably also on line 522 in case `scopeParam` is null and `clientScopes` are returned early.